### PR TITLE
[BEAM-7389] Add DoFn methods sample

### PR DESCRIPTION
**Pull request by davidcavazos**
Adding a code sample to demonstrate the use of the `setup`, `start_bundle`, `finish_bundle`, and `teardown` methods for a `DoFn`.

Also improved the captured output for the `dofn_params` sample.

R: @aaltay Can you give this a check whenever you have a chance? Thanks!

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py [hunk 1]
```diff
@@ -29,7 +29,8 @@ def __init__(self, delimiter=','):
       self.delimiter = delimiter
 
     def process(self, text):
-      return text.split(self.delimiter)
+      for word in text.split(self.delimiter):
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py]: Why this change?

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo_test.py [hunk 2]
```diff
@@ -58,24 +58,44 @@ def check_dofn_params(actual):
...
+* process: 🍓
+* process: 🥕
+* process: 🍆
+* process: 🍅
+* process: 🥔
+* finish_bundle: 🌱🌳🌍
+teardown
+[END results]'''.splitlines()[1:-1]
+  results = [line for line in results if line.startswith('*')]
+  assert_that(actual, equal_to(results))
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo_test.py]: Does this only validate that, process lines are in the output? Does it check for other lines?

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py (hunk 1 above)
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py]: I just thought it would be clearer to read and more consistent with other samples, although it's doing the same thing.

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo_test.py (hunk 2 above)
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo_test.py]: Good point. The sample code prints that as an output, and it makes sense to see what gets called after what. However, for the test we're only testing for the elements produced by `process` and `finish_bundle` since those are the ones that affect the output `PCollection`. So that's why I marked them with a `*` to make them easier to spot and test for. I'm adding a comment for this.

**Comment by davidcavazos** [comment]
@aaltay this is ready for review :)

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py [hunk 3]
```diff
@@ -81,3 +82,44 @@ def process(self, elem, timestamp=beam.DoFn.TimestampParam, window=beam.DoFn.Win
...
+  # [START pardo_dofn_methods]
+  import apache_beam as beam
+
+  class DoFnMethods(beam.DoFn):
+    def __init__(self):
+      print('__init__')
+      self.window = beam.window.GlobalWindow()
+
+    def setup(self):
+      print('setup')
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py]: I tried doing that here, but the code sample ended up looking a lot more cluttered and intimidating. I think it looked better in the docs themselves. Here's an extract of what I was adding in the docs:

A [`DoFn`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn) can be customized with a number of methods that can help create more complex behaviors. You can customize what a worker will do when it starts and shuts down with `setup` and `teardown`. You can also customize what to do when a [*bundle of elements*](https://beam.apache.org/documentation/execution-model/#bundling-and-persistence) starts and when a bundle finishes with `start_bundle` and `finish_bundle`.

* [`DoFn.setup()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.setup): Called *once per worker* when the worker is starting to run. This is a good place to connect to database instances, open network connections or other resources.

* [`DoFn.start_bundle()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.start_bundle): Called *once per bundle of elements* before calling `process` on the first element of the bundle. This is a good place to start keeping track of the bundle elements.

* [**`DoFn.process(element, *args, **kwargs)`**](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.process) *[required]:* Called *once per element*, can *yield zero or more elements*.

* [`DoFn.finish_bundle()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.finish_bundle): Called *once per bundle of elements* after calling `process` after the last element of the bundle, can *yield zero or more elements*. This is a good place to do batch calls on a bundle of elements, such as running a database query. For example, you can initialize a batch in `start_bundle`, add elements to the batch in `process` instead of yielding them, then running a batch query on those elements on `finish_bundle`, and yielding all the results.

  Note that yielded elements from `finish_bundle` must be of the type `apache_beam.utils.windowed_value.WindowedValue`. You will need to provide a timestamp as a unix timestamp, which you can get from the last processed element. You will also need to provide a window, which you can get from the last processed element like in the example below.

* [`DoFn.teardown()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.teardown): Called *once per worker* when the worker is shutting down. This is a good place to close database instances, close network connections or other resources.

[the remark this answers is not in the document]
Comment (aaltay) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py]: Sounds good to do it in the doc. I glanced over the comment here but did not carefully read it. Let's review it in another PR.

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py (hunk 1 above)
Comment (aaltay) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo.py]: Should we revert this?

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo_test.py (hunk 2 above)
Comment (aaltay) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo_test.py]: Ping on this comment?

Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/pardo_test.py]: I'm changing the test to actually check for both the order of the methods executed, as well as all the elements being present without ordering constraints.

Regarding if `DoFn` methods work on all runners, looking at #7994 I could only find that [BEAM-7885 `DoFn.setup` doesn't run for streaming jobs in the DirectRunner](https://issues.apache.org/jira/browse/BEAM-7885), and [BEAM-7340 `DoFn.teardown` metrics are lost](https://issues.apache.org/jira/browse/BEAM-7340). I'm adding those into comments.

I looked into Jira too, but that was all that I could find, are you aware of any other issues with other runners?

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by davidcavazos** [comment]
@aaltay Tests are passing. More detailed descriptions will be available in the docs.
